### PR TITLE
Fix registration socket removal on windows

### PR DESCRIPTION
**Pull request by kkmsft**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change

/kind bug

> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

Fix registration socket removal on windows.

Until the bug - https://github.com/golang/go/issues/33357 is fixed, os.stat wouldn't return the
right mode(socket) on windows. Hence deleting the file, without checking whether
it's a socket, on windows.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by msau42** [comment]
/ok-to-test
/assign @verult 
who has been reviewing PRs related to registration socket cleanup

File: cmd/csi-node-driver-registrar/node_register.go [hunk 1]
```diff
@@ -36,23 +36,32 @@ func nodeRegister(
...
-			os.Exit(1)
-		}
-	}
-	if err != nil && !os.IsNotExist(err) {
-		klog.Errorf("failed to stat the socket %s with error: %+v", socketPath, err)
-		os.Exit(1)
-	}
 
 	var oldmask int
 	if runtime.GOOS == "linux" {
```
Comment (ddebroy) [inline cmd/csi-node-driver-registrar/node_register.go]: Instead of the `if` here, how about wrapping up the logic in a function along the lines of `cleanupSocketFile()` and placing that in https://github.com/kubernetes-csi/node-driver-registrar/blob/master/cmd/csi-node-driver-registrar/util_linux.go and https://github.com/kubernetes-csi/node-driver-registrar/blob/master/cmd/csi-node-driver-registrar/util_windows.go ?

File: cmd/csi-node-driver-registrar/node_register.go [hunk 2]
```diff
@@ -36,23 +36,32 @@ func nodeRegister(
...
+		fi, err := os.Stat(socketPath)
+		if err == nil && (fi.Mode()&os.ModeSocket) != 0 {
```
Comment (ddebroy) [inline cmd/csi-node-driver-registrar/node_register.go]: Is the `(fi.Mode()&os.ModeSocket) != 0)` useful for Linux? I understand this is exiting code - so q for the maintainers.

File: cmd/csi-node-driver-registrar/node_register.go [hunk 3]
```diff
@@ -36,23 +36,32 @@ func nodeRegister(
...
+			os.Exit(1)
+		}
 		// Default to only user accessible socket, caller can open up later if desired
 		oldmask, _ = umask(0077)
+	} else {
+		// Until the bug - https://github.com/golang/go/issues/33357 is fixed, os.stat wouldn't return the
+		// right mode(socket) on windows. Hence deleting the file, without checking whether
+		// its a socket, on windows.
+		klog.Warningf("Windows environment detected. Deleting the socket path: %s", socketPath)
+		if err := os.Remove(socketPath); err != nil {
```
Comment (ddebroy) [inline cmd/csi-node-driver-registrar/node_register.go]: You do want to `stat` first for Windows and `os.Remove` only if the path exists, right?

Comment (kkmsft) [reply cmd/csi-node-driver-registrar/node_register.go]: yep, will modify.

File: cmd/csi-node-driver-registrar/node_register.go (hunk 1 above)
Comment (kkmsft) [reply cmd/csi-node-driver-registrar/node_register.go]: makes sense. Will update the PR

File: cmd/csi-node-driver-registrar/node_register.go [hunk 4]
```diff
@@ -36,23 +36,32 @@ func nodeRegister(
...
+				os.Exit(1)
+			}
+		}
+		if err != nil && !os.IsNotExist(err) {
+			klog.Errorf("failed to stat the socket %s with error: %+v", socketPath, err)
+			os.Exit(1)
+		}
 		// Default to only user accessible socket, caller can open up later if desired
 		oldmask, _ = umask(0077)
+	} else {
```
Comment (jingxu97) [inline cmd/csi-node-driver-registrar/node_register.go]: is it better to use 
if runtime.GOOS == "windows" {
...
} else {
...
}

File: cmd/csi-node-driver-registrar/node_register.go [hunk 5]
```diff
@@ -36,23 +36,32 @@ func nodeRegister(
...
+		// Until the bug - https://github.com/golang/go/issues/33357 is fixed, os.stat wouldn't return the
```
Comment (jingxu97) [inline cmd/csi-node-driver-registrar/node_register.go]: maybe put TODO: at the beginning for easy search later?

File: cmd/csi-node-driver-registrar/node_register.go (hunk 1 above)
Comment (kkmsft) [reply cmd/csi-node-driver-registrar/node_register.go]: Updated the PR. PTAL.

File: cmd/csi-node-driver-registrar/node_register.go (hunk 3 above)
Comment (kkmsft) [reply cmd/csi-node-driver-registrar/node_register.go]: Updated the PR. PTAL.

File: cmd/csi-node-driver-registrar/node_register.go (hunk 5 above)
Comment (kkmsft) [reply cmd/csi-node-driver-registrar/node_register.go]: Added the TODO in the util_windows.go file. PTAL.

File: cmd/csi-node-driver-registrar/node_register.go (hunk 4 above)
Comment (kkmsft) [reply cmd/csi-node-driver-registrar/node_register.go]: Logic moved to per os util file. so changes in this area. PTAL.

**Comment by ddebroy** [comment]
Thanks KK!
/lgtm

**Comment by kkmsft** [comment]
@verult - Please take a look.

File: cmd/csi-node-driver-registrar/node_register.go (hunk 1 above)
Comment (verult) [reply cmd/csi-node-driver-registrar/node_register.go]: I think Deep's intention was to not have to expose an OS check here. You could have a function where the Linux implementation calls umask() and the Windows implementation does nothing

File: pkg/util/util_linux.go [hunk 6]
```diff
@@ -16,12 +16,29 @@ See the License for the specific language governing permissions and
...
 )
 
-func umask(mask int) (int, error) {
+func Umask(mask int) (int, error) {
 	return unix.Umask(mask), nil
 }
+
+func CleanupSocketFile(socketPath string) error {
+	fi, err := os.Stat(socketPath)
+	if err == nil && (fi.Mode()&os.ModeSocket) != 0 {
```
Comment (verult) [inline pkg/util/util_linux.go]: FYI this might be buggy: https://github.com/kubernetes-csi/node-driver-registrar/pull/61#discussion_r377326338

We should return an error if the file isn't a socket. But yeah can be a separate PR

Comment (verult) [reply pkg/util/util_linux.go]: @Madhu-1 FYI. IMO this can be reused in the deletion goroutine. We don't *need* the socket check there but it doesn't hurt. If another file type is there, the socket is gone anyway.

File: pkg/util/util_test.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+	"k8s.io/klog"
+)
+
+func TestSocketPath(t *testing.T) {
+	// Create in the path where the test is run
+	socketPath := "reg.sock"
+	fileExists := false
+
+	// Test1: file does not exist. So clean up should be successful.
```
Comment (verult) [inline pkg/util/util_test.go]: Could you split each test case into its own function? Easier to see which has failed

File: pkg/util/util_test.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+	err := CleanupSocketFile(socketPath)
+	if err != nil {
+		t.Fatalf("cleanup returned error: %+v", err)
+	}
+
+	// Test2: director does not exist. So clean up should be successful.
```
Comment (verult) [inline pkg/util/util_test.go]: nit: `director` -> `directory`

File: pkg/util/util_linux.go (hunk 6 above)
Comment (kkmsft) [reply pkg/util/util_linux.go]: Thank you, @verult . Will raise a shorter PR with that change right after this one gets merged.

Comment (kkmsft) [reply pkg/util/util_linux.go]: For the umask will follow up with a short PR removing the checks, since that is existing logic being modified.

File: cmd/csi-node-driver-registrar/node_register.go (hunk 1 above)
Comment (kkmsft) [reply cmd/csi-node-driver-registrar/node_register.go]: For the umask will follow up with a short PR removing the checks, to separate out changes to existing code unrelated to this PR's context.

File: pkg/util/util_test.go (hunk 8 above)
Comment (kkmsft) [reply pkg/util/util_test.go]: Fixed.

File: pkg/util/util_test.go (hunk 7 above)
Comment (kkmsft) [reply pkg/util/util_test.go]: Changed to separate tests.

**Comment by kkmsft** [comment]
@verult, @ddebroy  - Addressed review comments, please take a final look. The lgtm labels got removed since I updated the PR.

I will follow up with two small PR:
1. Fixing the logic in the linux cleanup to return error if a non-socket file is found.
2. Remove the umask GOOS check in the code paths.

@msau42 - Addressed @verult and @ddebroy comments towards lgtm. Addressed comment raised by @jingxu97.  Please take a look and consider for merging.

File: pkg/util/util_windows.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
+// +build windows
+
+/*
+Copyright 2019 The Kubernetes Authors.
```
Comment (jingxu97) [inline pkg/util/util_windows.go]: 2020?

Comment (kkmsft) [reply pkg/util/util_windows.go]: Thank you, @jingxu97 - Updated to 2020.

File: pkg/util/util_test.go [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+	"net"
+	"os"
+	"path/filepath"
+	"runtime"
+	"testing"
+
+	"k8s.io/klog"
+)
+
+var socketPath = "reg.sock"
```
Comment (msau42) [inline pkg/util/util_test.go]: Generally we create a unique tmp directory per testcase and create test files inside there, so that we can run multiple unit tests in parallel.

Comment (kkmsft) [reply pkg/util/util_test.go]: @msau42 : Updated the PR to have tmp test directory. PTAL.

File: pkg/util/util_test.go [hunk 11]
```diff
@@ -17,51 +17,64 @@ limitations under the License.
...
 	"testing"
+	"time"
 
 	"k8s.io/klog"
 )
 
-var socketPath = "reg.sock"
+var socketFileName = "reg.sock"
+
+func getRandomTestDir() string {
```
Comment (msau42) [inline pkg/util/util_test.go]: How about using this helper? https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_plugin_test.go#L45

Comment (kkmsft) [reply pkg/util/util_test.go]: Thank you, @msau42 . Used the helper function and rebased the PR. PTAL.

**Comment by kkmsft** [comment]
@msau42 - have fixed the tests, go mod vendor etc. PTAL.

**Comment by msau42** [comment]
Thank you!
/lgtm
/approve
